File: cpp/day01/src/part2.cxx
#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>
#include <fstream>
#include <iostream>
#include <ranges>
#include "lib.hpp"

using namespace std;

int main(int argc, char **argv) {
  if (argc != 2) {
    cout << "Invalid input parameters. Usage: part1 <file path>" << endl;
  }
  char *filename = argv[1];

  ifstream is(filename);
  vector<int> left;
  std::unordered_map<int, int> right;

  for (string line; getline(is, line);) {
    auto items = line_to_numbers(line);
    left.push_back(get<0>(items));
    if (right.find(get<1>(items)) == right.end()) {
      right[get<1>(items)] = 1;
    } else {
      right[get<1>(items)] += 1;
    }
  }
  int result = 0;
  for (auto item : left) {
    try {
      int freq = right.at(item);
      result += freq * item;
    } catch(out_of_range) {
    }
  }
  cout << result << endl;
  return 0;
}

File: rust/day01/src/bin/part2.rs
use std::{collections::HashMap, env, fs};

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Require input file path");
    }

    let file_path = &args[1];
    let content = fs::read_to_string(file_path).expect("cannot read file");
    let mut left: Vec<i32> = vec![];
    let mut right: HashMap<i32, i32> = HashMap::new();
    for line in content.trim().split('\n') {
        let items: Vec<&str> = line.split(' ').collect();
        left.push(items.first().expect("invalid input").parse().unwrap());
        let right_value: i32 = items.last().expect("invalid input").parse().unwrap();
        *right.entry(right_value).or_insert(0) += 1;
    }
    let mut result: i32 = 0;
    for val in left.iter() {
        if let Some(freq) = right.get(val) {
            result += freq * val;
        }
    }
    println!("{result}");
}
